extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/ShardReader.java

```diff
@@ -0,0 +1,278 @@
+        }
+
+        return Result.NOTHING;
+    }
+
+    private Result handleWaitingForRecords() {
+        if (recordsResult.isDone()) {
+            try {
+                GetRecordsResult result = helper.readResult(recordsResult);
+                shardIterator = result.getNextShardIterator();
+                data.addAll(result.getRecords());
+                if (shardIterator == null) {
+                    state = State.SHARD_CLOSED;
+                    return data.size() > 0 ? Result.HAS_DATA : Result.CLOSED;
+                } else if (data.size() > 0) {
+                    state = State.HAS_DATA_NEED_TO_REQUEST_RECORDS;
+                    return Result.HAS_DATA;
+                } else {
+                    state = State.NEED_TO_REQUEST_RECORDS;
+                    return Result.NOTHING;
+                }
+            } catch (ProvisionedThroughputExceededException pte) {
+                return dealWithReadRecordFailure("Data throughput rate exceeded. Backing off and retrying.");
+            } catch (ExpiredIteratorException eie) {
+                return dealWithReadRecordFailure("Record iterator expired. Retrying.");
+            } catch (SdkClientException sce) {
+                return dealWithReadRecordFailure("Failed reading records, retrying. Cause: " + sce.getMessage());
+            } catch (Throwable t) {
+                throw rethrow(t);
+            }
+        } else {
+            return Result.NOTHING;
+        }
+    }
+
+    private Result dealWithReadRecordFailure(String message) {
+        logger.warning(message);
+        nextGetRecordsTime = System.nanoTime() + PAUSE_AFTER_FAILURE;
+        state = State.NEED_TO_REQUEST_RECORDS;
+        return Result.NOTHING;
+    }
+
+    private Result handleHasDataNeedToRequestRecords() {
+        if (attemptToSendGetRecordsRequest()) {
+            state = data.size() > 0 ? State.HAS_DATA : State.WAITING_FOR_RECORDS;
+        }
+
+        return data.size() > 0 ? Result.HAS_DATA : Result.NOTHING;
+    }
+
+    private Result handleHasData() {
+        state = data.size() > 0 ? State.HAS_DATA : State.WAITING_FOR_RECORDS;
+        return data.size() > 0 ? Result.HAS_DATA : Result.NOTHING;
+    }
+
+    private Result handleShardClosed() {
+        return data.size() > 0 ? Result.HAS_DATA : Result.CLOSED;
+    }
+
+    private boolean attemptToSendGetRecordsRequest() {
+        if (System.nanoTime() < nextGetRecordsTime) {
+            return false;
+        }
+
+        recordsResult = helper.getRecordsAsync(shardIterator);
+
+        nextGetRecordsTime = System.nanoTime() + getRecordsRateTracker.next();
+        return true;
+    }
+
+    public Shard getShard() {
+        return shard;
+    }
+
+    public Record[] getData() {
+        if (data.isEmpty()) {
+            throw new IllegalStateException("Can't ask for data when none is available");
+        }
+
+        Record[] records = new Record[Math.min(data.size(), DATA_BATCH_SIZE)];
```

Comment:
There's a bit too much copying: we copy the results from `GetRecordsResult` to `LinkedList<Record>` in `handleWaitingForRecords`. Then from the linked list to an array in `getData()`. I think neither of this is necessary - can't we directly return the `List<Record>` from the `GetRecordsResult` in `getData`? Why do we need the batching here?

Reply:
I've added the batching because `GetRecordsResult` can contain up to 10,000 records and since the source is supposed to be cooperative, that's a bit much to return in one go, or is it not?

Reply:
We should add to the outbox as much as the outbox accepts. We should limit the time spent in the call, but we're just copying items so we're not doing much work. This is what we do elsewhere: the outbox free capacity is the limit for the batch size.